---
title: "Methylation analysis of metabolomically-driven tumours"
author: "Philip Smith"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# Introduction
## Preface
### Resources

- https://bioconductor.org/packages/release/bioc/vignettes/minfi/inst/doc/minfi.html#8_identifying_differentially_methylated_regions
- https://academic.oup.com/bioinformatics/article/33/4/558/2666344
- https://www.ncbi.nlm.nih.gov/pubmed/29469594
- https://www.bioconductor.org/help/course-materials/2015/BioC2015/methylation450k.html#identifying-dmrs-and-dmps
- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6366976/
- https://www.ncbi.nlm.nih.gov/pubmed/23314698?dopt=Abstract
- http://www.epicenteredresearch.com/amlbinder/SBE2015/Day2/RMarkdown_for_RA_final.html


## Env Setup
### Install bioc manager
```{r install_bioc,warning=FALSE}
suppressPackageStartupMessages(
if (!requireNamespace("BiocManager", quietly = TRUE)){
  install.packages("BiocManager",dependencies = TRUE)
  library(BiocManager)
}
)
```

### CRAN packages
```{r install_cran,warning=FALSE}
CRANpackages <- c("tidyverse","reshape2","ggplot2","cowplot","ggsignif",
                  "circlize","RColorBrewer","ggdendro")
suppressPackageStartupMessages(
for(package in CRANpackages){
  if (!sapply(package,require,character=TRUE)){
    install.packages(package,dependencies = TRUE)
    sapply(package,require,character=TRUE)

  }
}
)
```

### Bioc packages
```{r install_biocP,warning=FALSE}
bicopackages <- c("IlluminaHumanMethylationEPICmanifest","DMRcate",
                  "IlluminaHumanMethylationEPICanno.ilm10b4.hg19",
                  "GenomicFeatures","TxDb.Hsapiens.UCSC.hg19.knownGene",
                  "org.Hs.eg.db","minfi","sva","doParallel","topGO","limma")
suppressPackageStartupMessages(
for(package in bicopackages){
  if (!sapply(package,require,character=TRUE)){
    BiocManager::install(package)
    sapply(package,require,character=TRUE)
  }
}
)
```

### Functions
```{r funcs}
## Function for performing data QC
source("resources/funcs.R")
```

# Analysis
## Analysis setup
### Required variables
```{r vars}
## Vars
preprocess.var <- "Funnorm"
samplesheet <- "data/850K_final_sample_sheet_recoded_NOV2019_FINAL.txt"
excluded <- c("D1234004","s065669","S164733") # ALWAYS EXCLUDE
excluded <- c(excluded,"TB17.0821","gtb81831") # ADD ADDITIONAL SAMPLES HERE
test_group <- c("PPGL","GLIOMA")
cores <- 4
bmphunter_cutoff <- 0.3
bmphunter_permutation <- 1000
test_pheno <- "Mutation_status"
dmr_distance <- 2000
SVA  <- FALSE

# Init log files
write.table(x = paste("Sample","Type",sep = "\t"),
            file = paste("excluded_samples_",paste(test_group,collapse = "_"),".txt",sep = ""),
            sep = "\t",
            append = F,
            col.names = F,
            row.names = F,
            quote = F)

write.table(x = paste("Probe","Type",sep = "\t"),
                file = paste("excluded_probes_",paste(test_group,collapse = "_"),".txt",sep = ""),
                sep = "\t",
                append = T,
                col.names = F,
                row.names = F,
                quote = F)
```

### Load data
```{r load_sample}
## Load sample sheet & exclude known unwanted samples
sample_sheet <- read.table(file = samplesheet,header = T,stringsAsFactors = F,sep = "\t",strip.white = T)
## Report sample sheet errors resulting in exclusions
write.table(x = paste(sample_sheet$Sample_Name[sample_sheet$Exclude == TRUE],"SAMPLE_SHEET_ERRORS",sep = "\t"),
            file = paste("excluded_samples_",paste(test_group,collapse = "_"),".txt",sep = ""),
            sep = "\t",
            append = T,
            col.names = F,
            row.names = F,
            quote = F)
## Final sample sheet
sample_sheet <- sample_sheet[sample_sheet$Exclude == FALSE,]
head(sample_sheet)
```

```{r load_ann}
## Manifest ann
annEPIC <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
```

```{r load_methylation,warning=FALSE}
## Read methylation data & label
Meth.data <- read.metharray.exp(targets = sample_sheet)
sampleNames(Meth.data) <- as.character(sample_sheet$Sample_Name)
```

### First pass QC
#### Manually excluded samples

```{r man_exclude}
excluded
## Report excluded samples
write.table(x = paste(excluded,"MANUAL",sep = "\t"),
            file = paste("excluded_samples_",paste(test_group,collapse = "_"),".txt",sep = ""),
            sep = "\t",
            append = T,
            col.names = F,
            row.names = F,
            quote = F)
```

Remove manually excluded samples

```{r exclude_man_methyl}
## Manual exclusion of specific samples
Meth.data <- Meth.data[,!colnames(Meth.data) %in% excluded]
```

#### Sex assignment
```{r sex_qc}
sex.qc(Meth.data)
```

#### Probe-level array
```{r probe_qc}
Meth.data.QC <- probe.QC(Meth.data)
```

```{r probe_qc_plot,fig.width=12}
## General plots
col.probe.p <- colMeans(detectionP(Meth.data))
ggplot() + geom_col(aes(y = col.probe.p,x = names(col.probe.p)),color = "grey15",fill="cadetblue3") +
  geom_hline(yintercept = 0.05,linetype=2) +
  geom_hline(yintercept = 0.01,linetype=3) +
  labs(title = "Methylation Detection significance - Cummulative per sample") +
  xlab("Sample") +
  ylab("ColMean(detectP-function)") +
  scale_y_continuous(expand = c(0,0),limits = c(0,0.25)) +
  theme(axis.text.x = element_text(angle = 90))
```

#### Sample-level array
```{r sample_qc}
Meth.data.QC <- sample.QC(Meth.data.QC)
```

### Normalisation
```{r norm_array}
## Perform normalisation method
gRS <- normalise(exp.data = Meth.data.QC)
```

#### Remove X/Y
Mixed sex sample sets are complicated by chrX and chrY probes so are dropped in this instance.
```{r sex_drop}
## Drop chrX and chrY probes - mixed sex subset
write.table(x = paste(rownames(gRS[featureNames(gRS) %in% annEPIC$Name[annEPIC$chr %in% c("chrX","chrY")],]),"CHR_X/Y",sep = "\t"),
            file = paste("excluded_probes_",paste(test_group,collapse = "_"),".txt",sep = ""),
            sep = "\t",
            append = T,
            col.names = F,
            row.names = F,
            quote = F)
gRS.flt <- gRS[!featureNames(gRS) %in% annEPIC$Name[annEPIC$chr %in% c("chrX","chrY")],]
```

#### SNP-Probe overlaps
```{r snp_prove_filt}
## Filter probes overlaping with common SNPs
write.table(x = paste(rownames(gRS.flt)[!rownames(gRS.flt) %in% rownames(dropLociWithSnps(gRS.flt, snps=c("SBE","CpG"), maf=0.05))],"SNP_MAF_5pct",sep = "\t"),
            file = paste("excluded_probes_",paste(test_group,collapse = "_"),".txt",sep = ""),
            sep = "\t",
            append = T,
            col.names = F,
            row.names = F,
            quote = F)
gRS.flt <- dropLociWithSnps(gRS.flt, snps=c("SBE","CpG"), maf=0.05)
```

#### Cross-reactivity
```{r x-react}
## Cross-reactivity data - as provided here https://doi.org/10.1186/s13059-016-1066-1
x.reactive <- read.table("resources/13059_2016_1066_MOESM1_ESM.csv",sep = ",",header = T,stringsAsFactors = F)

## Dropping cross reactive probes
write.table(x = paste(rownames(gRS.flt[featureNames(gRS.flt) %in% x.reactive$Name,]),"cross_reactive",sep = "\t"),
            file = paste("excluded_probes_",paste(test_group,collapse = "_"),".txt",sep = ""),
            sep = "\t",
            append = T,
            col.names = F,
            row.names = F,
            quote = F)
gRS.flt <- gRS.flt[!featureNames(gRS.flt) %in% x.reactive$Name,]
```

#### Array QC plot

```{r norm_plot,fig.width=12}
## Density Plots -before and after normalisation and QC
par(mfrow=c(1,3))
densityPlot(Meth.data.QC,main = "No normalisation")
densityPlot(getBeta(gRS),main = "preprocessFunnorm normalisation")
densityPlot(getBeta(gRS.flt),main = "Post-normalised filtering")
```

## Analysis
#### Array clustering
```{r cluster_array}
## Clustering all data
clrs.clst <- brewer.pal(nlevels(as.factor(pData(gRS.flt)$Group)),"Dark2")[as.factor(pData(gRS.flt)$Group)]
names(clrs.clst) <- as.factor(pData(gRS.flt)$Group)

pca.clst <- t(getBeta(gRS.flt))
pca.clst.prc <- as.data.frame(prcomp(pca.clst,scale. = T, center = T)$x)
histCLUST <- as.dendrogram(hclust(dist(scale(pca.clst),method = "euclidean",diag = F),method = "ward.D2"))
histCLUST <- dendrapply(histCLUST,colLab)
```

```{r plot_clust_hist,fig.width=12}
plot(histCLUST, type = "triangle", ylab = "")
legend("topright",
       legend = unique(names(clrs.clst)),
       col = unique(clrs.clst),
       pch = 15, bty = "n",  pt.cex = 1, cex = 0.5,
       text.col = "black", horiz = FALSE)
```

```{r plot_clust_pca,fig.width=12}
ggplot(as.data.frame(pca.clst.prc)) +
  geom_point(aes(PC1,PC2,colour = as.factor(pData(gRS.flt)$Group),shape = as.factor(pData(gRS.flt)$Mutation_status))) +
  theme_minimal() +
  theme(legend.direction = "horizontal",legend.position = "bottom",axis.line = element_line()) +
  scale_color_discrete(name = "Group") +
  scale_shape_discrete(name = "Mutation status", labels = c("WT","MUT"))
```

#### Sample sheet summary
```{r sample_summary}
## Summary plot and stats
sample_sheet$split <- paste(sample_sheet$Group,ifelse(sample_sheet$Mutation_status == 1,"mutated","non-mutated"),sep = " ")
splits <- split(x = sample_sheet[,1],f = factor(sample_sheet$split))
meanBetaPerGroup <- lapply(X = splits,FUN = function(x){colMeans(getBeta(gRS.flt[,colnames(gRS.flt) %in% x]),na.rm = T)})
```

#### Global methylation
```{r median_methylation,fig.width=12}
## Non-parametric
kwt <- kruskal.test(value ~ L1,data = melt(meanBetaPerGroup))
wilcox <- pairwise.wilcox.test(melt(meanBetaPerGroup)$value,melt(meanBetaPerGroup)$L1,p.adjust = "none")

print(kwt)
print(wilcox)

ggplot(melt(meanBetaPerGroup)) +
  geom_point(aes(y=value,x=L1,color = L1),position="jitter") +
  geom_boxplot(aes(x=L1,y=value,fill= L1)) +
  geom_signif(aes(x=L1,y=value),
              comparisons = list(c("GIST mutated","GIST non-mutated"),
                                 c("PPGL mutated","PPGL non-mutated")),
              test = "wilcox.test",y_position = 0.7,map_signif_level = F) +
  scale_y_continuous(limits = c(NA,0.85)) +
  scale_fill_manual(name="Group",
                    values = alpha(c("darkolivegreen1","darkolivegreen3","gold1","gold3"),0.7),
                    aesthetics = c("color","fill")) +
  ylab("Mean β value") +
  xlab(NULL) +
  labs(title = "Mean methylation per group") +
  theme_bw() +
  theme(legend.position = "bottom",
        legend.direction = "horizontal")
```

```{r density_methylation,fig.width=12}
ggplot(melt(meanBetaPerGroup)) +
  geom_density(aes(value, fill=L1)) +
  scale_fill_manual(name="Group",
                    values = alpha(c("darkolivegreen1","darkolivegreen3","gold1","gold3"),0.7)) +
  scale_x_continuous(limits =  c(0,1)) +
  ylab("Density") +
  xlab(NULL) +
  theme_bw() +
  theme(legend.position = "bottom",
        legend.direction = "horizontal") +
  facet_wrap(.~L1,nrow = 3,ncol=2)
```

#### Save filtered data
```{r save}
# TEMP Save filtered methyl data for downstream analysis
saveRDS(gRS.flt,file = "filtered.methyl.data.rds")
```

#### KMD6A/B/C

```{r get_betas}
methylBeta <- getBeta(gRS.flt)
methylBeta <- as.data.frame(methylBeta) %>%
  rownames_to_column(var = "probe")
```

```{r gene_ranges}
KDM4_gene_ranges <- GRanges(seqnames = c("chr1", "chr9", "chr19"), strand = c("+","+","+"),
               ranges = IRanges(start = c(43650148, 6720863, 4969113),
                                end = c(43705517,7175648,5153598),
                                names = c("KDM4A","KDM4C","KDM4B")))

probe_grange <- makeGRangesFromDataFrame(annEPIC,start.field = "pos",end.field = "pos")

probe_KDM4_overlaps <- as.data.frame(subsetByOverlaps(probe_grange,KDM4_gene_ranges))

KDM_methyl <- probe_KDM4_overlaps %>%
      rownames_to_column(var = "probe") %>%
      inner_join(.,methylBeta,by="probe") %>%
      pivot_longer(cols = 7:ncol(.))

KDM_methyl$Group <- sample_sheet$Group[match(KDM_methyl$name,sample_sheet$Sample_Name)]
KDM_methyl$Mut_status <- sample_sheet$Mutation_status[match(KDM_methyl$name,sample_sheet$Sample_Name)]
KDM_methyl$gene <- rownames(as.data.frame(KDM4_gene_ranges))[match(KDM_methyl$seqnames,as.data.frame(KDM4_gene_ranges)$seqnames)]

KDM_methyl <- KDM_methyl %>%
                dplyr::select(-probe,-seqnames,-end,-width,-strand) %>%
                group_by(start,Group,Mut_status,gene) %>%
                summarise(beta_median=median(value))


diff <- KDM_methyl$beta_median[KDM_methyl$Mut_status == "1"] - KDM_methyl$beta_median[KDM_methyl$Mut_status == "0"] 
KDM_methyl_diff <- KDM_methyl[KDM_methyl$Mut_status == "1",c("start","Group","gene")]
KDM_methyl_diff$diff <- diff

ggplot(KDM_methyl_diff) +
  geom_point(aes(x = start,y = diff)) +
  geom_hline(yintercept = 0) +
  scale_y_continuous(limits = c(-1,1)) +
  labs(title = "Relative differene in probe-level Beta between w/t and mut groups",
       subtitle = "Positive value is increased methylation / Negative value is decreased methylation") +
  facet_grid(Group ~ gene,scales = "free")

```

### Metabolomics

```{r groups}
group <- c("PPGL","GLIOMA")
```

```{r metabolic corr}
## correlation/Distribution - metabolomic - methylation
metabolomic <- sample_sheet[which(colnames(sample_sheet) %in% c("Sample_Name","X2HG_concentration_uM","Succinate_concentration_uM","Group","Mutation_status"))]
mean_methyl <- data.frame(mean_methyl=colMeans(getBeta(gRS.flt),na.rm = T))
meta.methyl.merge <- merge(mean_methyl,metabolomic,by.x=0,by.y="Sample_Name",all.y=T)
```

```{r metabolic_plots}
meth.dist <- ggplot(meta.methyl.merge) +
  geom_histogram(aes(mean_methyl),bins = 10,fill="springgreen",color="grey15") +
  xlab(label = "mean β value") +
  theme_linedraw() +
  theme(axis.title.y = element_blank(),plot.margin = unit(x = c(0.01,0.1,0.01,0.01),units = "npc"))

HG.dist <- ggplot(meta.methyl.merge) +
  geom_histogram(aes(X2HG_concentration_uM),bins = 10,fill="tomato",color="grey15") +
  xlab(label = "2HG concentration (µM)") +
  theme_linedraw() +
  theme(axis.title.y = element_blank(),plot.margin = unit(x = c(0.01,0.1,0.01,0.01),units = "npc"))

succ.dist <- ggplot(meta.methyl.merge) +
  geom_histogram(aes(Succinate_concentration_uM),bins = 5,fill="skyblue",color="grey15") +
  xlab(label = "Succinate concentration (µM)") +
  theme_linedraw() +
  theme(axis.title.y = element_blank(),plot.margin = unit(x = c(0.01,0.1,0.01,0.01),units = "npc"))

X2HG.data <- meta.methyl.merge[meta.methyl.merge$Group %in% "GLIOMA" & !is.na(meta.methyl.merge$X2HG_concentration_uM) & !is.na(meta.methyl.merge$mean_methyl),]
Succinate.data <- meta.methyl.merge[meta.methyl.merge$Group == "PPGL" & !is.na(meta.methyl.merge$Succinate_concentration_uM) & !is.na(meta.methyl.merge$mean_methyl),]

HG <- ggplot(X2HG.data) +
  geom_point(aes(x = mean_methyl,y = X2HG_concentration_uM,color=as.factor(Mutation_status))) +
  geom_smooth(aes(mean_methyl,X2HG_concentration_uM),method = "lm",se = F) +
  xlab("mean β value") + ylab("2HG concentration (µM)") +
  theme_linedraw() +
  theme(legend.position = "none",plot.margin = unit(x = c(0.01,0.1,0.01,0.01),units = "npc"))
HG <- HG + annotate(geom = "text",x = 0.57,y = 5200,hjust = 0,label = paste("p.value =",
                                                                            signif(cor.test(x = X2HG.data$mean_methyl,y = X2HG.data$X2HG_concentration_uM,method = "spearman")$p.value,digits = 3))) +
  annotate(geom = "text",x = 0.57,y = 5000,hjust = 0,label = paste("spearman rho =",
                                                                   signif(cor.test(x = X2HG.data$mean_methyl,y = X2HG.data$X2HG_concentration_uM,method = "spearman")$estimate,digits = 3)))

succ <- ggplot(Succinate.data) +
  geom_point(aes(x = mean_methyl,y = Succinate_concentration_uM,color=as.factor(Mutation_status))) +
  geom_smooth(aes(mean_methyl,Succinate_concentration_uM),method = "lm",se = F) +
  xlab("mean β value") + ylab("Succinate concentration (µM)") +
  theme_linedraw() +
  theme(legend.position = "none",plot.margin = unit(x = c(0.01,0.1,0.01,0.01),units = "npc"))
succ <- succ + annotate(geom = "text",x = 0.45,y = 7000,hjust = 0,label = paste("p.value =",
                                                                                signif(cor.test(x = Succinate.data$mean_methyl,y = Succinate.data$Succinate_concentration_uM,method = "spearman")$p.value,digits = 3))) +
  annotate(geom = "text",x = 0.45,y = 6800,hjust = 0,label = paste("spearman rho =",
                                                                   signif(cor.test(x = Succinate.data$mean_methyl,y = Succinate.data$Succinate_concentration_uM,method = "spearman")$estimate,digits = 3)))

plot_grid(plotlist = list(plot_grid(meth.dist,succ.dist,HG.dist,ncol = 3,labels = c("A","B","C")),plot_grid(HG,succ,labels = c("D","E"))),nrow = 2)
```

### In progress

This section is being worked on but no results are ready.
This should provide a statistical frame work to define if KDM genes are differentially methylated.

#### DMR modelling

##### PPGL
```{r subset_methyl_group}
## Subset data by samples of interest defined by test_group var
sgRS <- gRS.flt[,pData(gRS.flt)$Group == "PPGL"]
```

```{r dmr_model}
## MODEL
mut_status <- pData(sgRS)[,colnames(pData(sgRS)) == "Mutation_status"]
tumour_group <- pData(sgRS)[,colnames(pData(sgRS)) == "Group"]
age <- pData(sgRS)[,colnames(pData(sgRS)) == "Age"]
design <- model.matrix(~ mut_status)
```

```{r DMP_sig}
## DMP
beta <- getBeta(sgRS)
dmp <- dmpFinder(beta, pheno = mut_status,type = "categorical",qCutoff = 0.1)
write.table(x = dmp[dmp$pval < 0.05,],
            file = paste("DMP_table_",paste("PPGL",collapse = "_"),".tsv",sep = ""),
            quote = F,sep = "\t",row.names = T,col.names = T)
```

```{r dmr_fitting}
## SVA implementation
if(SVA == TRUE){
  # mval <- getM(sgRS)
  # mval <- mval[-which(!is.finite(rowSums(mval))),]
  # mut_status.sva <- pData(sgRS)
  # 
  # mod <- model.matrix(~as.factor(Mutation_status), data=mut_status.sva)
  # mod0 <- model.matrix(~1, data=mut_status.sva)
  # 
  # sva.results <- sva(mval, mod, mod0)
  # modSv = cbind(mod,sva.results$sv)
  # fit <- lmFit(mval,modSv)
  # write.table(x = fit$design,file = paste("sva_bmpHunter_",paste(test_group,collapse = "_"),"_model.txt",sep = ""),quote = F,sep = "\t",row.names = T,col.names = T)
  # 
  # registerDoParallel(cores = cores)
  # dmrs <- bumphunter(sgRS, design = fit$design,cutoff = bmphunter_cutoff, B=bmphunter_permutation, type="Beta",nullMethod="bootstrap")
} else {
  # ## CHECK FOR NUMBER OF BUMPS AND ADJUST CUTOFF according - No SVA implementation
  # ## bumphunter
  # registerDoParallel(cores = 1)
  # dmrs.len <- bumphunter(sgRS, design = design,cutoff = bmphunter_cutoff, B=0, type="Beta")
  # nrow(dmrs.len$table)
  # dmrs <- bumphunter(sgRS, design = design,cutoff = bmphunter_cutoff, B=bmphunter_permutation, type="Beta")
}
```

```{r DMR_annotation}
## DMR annotation
# dmrs.GRange <- annotate_dmrs(x = dmrs)
# write.table(x = as.data.frame(dmrs.GRange),file = paste("dmrs_table_",paste(test_group,collapse = "_"),".tsv",sep = ""),append = F,quote = F,sep = "\t",row.names = F,col.names = T)
```

```{r DMR_GO_analysis}
## GO analysis
# GO.ids <- mapIds(x = org.Hs.eg.db,keys = as.character(dmrs.GRange$name),keytype = "SYMBOL",column = "GO",multiVals = "list")
# geneL <- dmrs.GRange$fwer
# names(geneL) <- as.character(dmrs.GRange$name)
# GOdata <- new("topGOdata",
#               ontology = "BP",
#               allGenes = geneL,
#               geneSel = geneSelction,
#               annot = annFUN.gene2GO,
#               gene2GO = GO.ids,
#               nodeSize = 5)
# 
# resultFisher <- runTest(GOdata, algorithm = "classic", statistic = "fisher")
# resultKS <- runTest(GOdata, algorithm = "classic", statistic = "ks")
# resultKS.elim <- runTest(GOdata, algorithm = "elim", statistic = "ks")
# resultKS.weight01 <- runTest(GOdata, algorithm = "weight01", statistic = "ks")
# 
# allRes <- GenTable(GOdata,Fisher = resultFisher,weight01KS = resultKS.weight01,KS = resultKS,KSelim = resultKS.elim,topNodes = 500,numChar = 500)
# write.table(x = as.data.frame(allRes),file = paste("GO_table_",paste(test_group,collapse = "_"),".tsv",sep = ""),append = F,quote = F,sep = "\t",row.names = F,col.names = T)
```

## Session info
```{r sessionInfo}
sessionInfo()
```

```{r end}
# END
```
